peterpy/handlers/products.py

```diff
@@ -89,12 +76,42 @@ async def add_product(request: Request) -> Response:
     )
 
 
+# Add this to show adding batch of products, while only
+# committing at the end of the batch in the middleware
+@routes.post("/products")
+async def add_products(request: Request) -> Response:
+    logging.debug("---------------------------------")
+    logging.info("Add multiple products requested from %s", request.remote)
+
+    product_service: ProductService = request["product_service"]
+
+    data = await request.json()
+    products = data.get("products")
+
+    try:
+        for product in products:
+            name = product.get("name")
+            price = product.get("price")
+            product = await product_service.add(name, price)
+
+        return json_response(
+            status=201,
+            text=json.dumps(
+                {"products": products},
+                cls=ProductEncoder,
+            ),
+        )
+    except ValueError as e:
+        return json_response(status=400, text=json.dumps({"error": str(e)}))
```

Comment:
catching the exception here to send a json response, but now the exception doesn't bubble up to the middleware, where the rollback now also doesn't happen